### ISSUE TITLE
Inconsistent Results in Dynamic Relaxation Module with Varying n_chunks"

### ISSUE THREAD

**Issue by ShiWeiHuH** [kaipartmann/Peridynamics.jl#187]
Julia Version 1.10.3 (2024-04-30)
Peridynamics Version v0.3.3

While testing the Dynamic Relaxation module, I noticed that using different n_chunks produces inconsistent results, whereas the Velocity Verlet module runs correctly. This suggests there might be data contention during parallel processing. I suspect the issue lies in the calculation of the density_matrix, which requires not only local node data but also neighboring node data. However, even when I calculate it without using multithreading, the results still vary with n_chunks. I now believe the problem might be related to the indexing in the relaxation_step calculation, and I’m currently working on modifying that.

**Comment by ShiWeiHuH** [comment]
Here is my code:
```julia
using Peridynamics
l = 1.0
w = 0.1
Δx = l/100
b = 1
dt = 1.0
nt = 4000
f = 2.0e10

pos, vol = uniform_box(l+3Δx, w, w, Δx; center_x=-1.5Δx, center_y=0, center_z=0)
body = Body(BBMaterial{}(), pos, vol) #EnergySurfaceCorrection
material!(body; horizon=3.015Δx, E=2.0e11, rho=7.85e3, epsilon_c=b)

point_set!(p -> (p[1] >= l/2 - Δx), body, :force_area)
point_set!(p -> (p[1] <= -l/2), body, :fixed_area)

forcedensity_bc!(t -> f, body, :force_area, :y)
velocity_bc!(t -> 0.0, body, :fixed_area, :x)
velocity_bc!(t -> 0.0, body, :fixed_area, :y)
velocity_bc!(t -> 0.0, body, :fixed_area, :z)

vv = DynamicRelaxation(steps=nt)
job = Job(body, vv; freq=10, path="results/static_ori_no_4")
A = submit(job)

**Comment by kaipartmann** [comment]
Hello @ShiWeiHuH, thank you very much for pointing out this serious issue. I was able to reproduce the bug.

I am currently investigating how to fix it, but I haven't found a solution yet. I will ping you as soon as I have a fix.

**Comment by kaipartmann** [comment]
@ShiWeiHuH How did you check the consistency of the results?
I encountered an error of the displacement in the order of ~1e-6 after all the time steps. So no visible difference in the post-processing with ParaView. I also could not find anything that indicates a bug in the halo-exchange. Maybe it is a (very large) numerical error due to the calculation of the dampening?

**Comment by ShiWeiHuH** [comment]
> @ShiWeiHuH How did you check the consistency of the results? I encountered an error of the displacement in the order of ~1e-6 after all the time steps. So no visible difference in the post-processing with ParaView. I also could not find anything that indicates a bug in the halo-exchange. Maybe it is a (very large) numerical error due to the calculation of the dampening?

**Comment by ShiWeiHuH** [comment]
I have been closely inspecting the results from parallel computations and noticed small numerical differences when running with different core counts. Although these discrepancies are minor, I am unsure if they indicate a potential issue with the computational logic.

I encountered a similar issue in version 0.2, which was due to the addition of new boundary conditions, and I was able to resolve it. Since this is a sensitive area, I wanted to ensure everything is correct before proceeding further.

My current task is to migrate several features to version 0.3, including heat transfer, thermomechanical coupling, and multi-time stepping. As part of this migration, I tested the parallel computations and noticed these small variations.
